r/R/dplyr-funcs-datetime.R

```diff
@@ -686,4 +686,106 @@ register_bindings_datetime_parsers <- function() {
+          options = cast_options(to_type = duration(unit = "s"))
+        )
+        interim <- build_expr("round_temporal", x - week_start_offset, options = opts)
+        return(interim + week_start_offset)
+      }
+    }
+    Expression$create("round_temporal", x, options = opts)
+  })
+
+  register_binding("floor_date", function(x, unit = "second",
+                                          week_start = getOption("lubridate.week.start", 7)) {
+    opts <- parse_period_unit(unit)
+    if (opts$unit == 7L) {
+      if (week_start == 7) { # Sunday
+        opts$week_starts_monday <- 0L
+        return(Expression$create("floor_temporal", x, options = opts))
+
+      } else if (week_start == 1) { # Monday
+        opts$week_starts_monday <- 1L
+        return(Expression$create("floor_temporal", x, options = opts))
+
+      } else { # for other values of week_start, compute using an offset
+        week_start_offset <- build_expr(
+          "cast",
+          Scalar$create((as.integer(week_start) - 1L) * 86400L, int64()),
+          options = cast_options(to_type = duration(unit = "s"))
+        )
+        interim <- build_expr("floor_temporal", x - week_start_offset, options = opts)
+        return(interim + week_start_offset)
+      }
+    }
+    Expression$create("floor_temporal", x, options = opts)
+  })
+
+  register_binding("ceiling_date", function(x, unit = "second",
+                                            change_on_boundary = NULL,
+                                            week_start = getOption("lubridate.week.start", 7)) {
+
+    opts <- parse_period_unit(unit)
+
+    if (is.null(change_on_boundary)) {
+      if (call_binding("is.Date", x)) {
+        change_on_boundary <- TRUE
+      } else {
+        change_on_boundary <- FALSE
+      }
+    }
+
+    if (change_on_boundary == FALSE) {
+      opts$ceil_is_strictly_greater <- 0L
+    }
+
+    if (change_on_boundary == TRUE) {
+      opts$ceil_is_strictly_greater <- 1L
+    }
+
+    if (opts$unit == 7L) {
+      if (week_start == 7) { # Sunday
+        opts$week_starts_monday <- 0L
+        return(Expression$create("ceil_temporal", x, options = opts))
+
+      } else if (week_start == 1) { # Monday
+        opts$week_starts_monday <- 1L
+        return(Expression$create("ceil_temporal", x, options = opts))
+
+      } else { # for other values of week_start, compute using an offset
+        week_start_offset <- build_expr(
+          "cast",
+          Scalar$create((as.integer(week_start) - 1L) * 86400L, int64()),
+          options = cast_options(to_type = duration(unit = "s"))
+        )
+        interim <- build_expr("ceil_temporal", x - week_start_offset, options = opts)
+        return(interim + week_start_offset)
+      }
+    }
+    return(Expression$create("ceil_temporal", x, options = opts))
+  })
+
+  register_binding("leap_year", function(date) {
+    year <- Expression$create("year", date)
+    (year %% 4 == 0) & ((year %% 100 != 0) | (year %% 400 == 0))
+  })
```

Comment:
I think this is solved with a c++ kernel now and can be removed? I might have reintroduced it with my rebase.